### DISCUSSION TITLE
Questions/suggestions about folder sync UX

### DISCUSSION THREAD

**Discussion by willurmston** [beakerbrowser/beaker#1723] [General] [answered]
As I excitedly tried out the new features in prerelease.4, I accidentally deleted the contents of both a drive and another drive mounted to `/.ui`. I was trying to sync a drive to a folder, so I created a new folder using the folder sync dialog

<img width="424" alt="image" src="https://user-images.githubusercontent.com/8092948/83831757-ed9fc000-a69c-11ea-9b76-ec05c80a9a7b.png">

When this dialog came up, I clicked the `Autosync` toggle, and the dialog closed and all the files in the drives were gone. Luckily I had the site open and cached in another tab so I could download the individual files.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/8092948/83831597-9863ae80-a69c-11ea-97e4-a7365fb37ba6.png">

1. Is that the intended behavior? I would expect all the files to be copied to the folder, and nothing destructive would happen.
2. If so, what is the proper way to start autosyncing a drive to an empty folder without deleting everything in the drive?
3. When you sync a drive to a folder, are mounted sub-drives also copied to the folder?

Thank you, these latest features are exactly what I need for what I'm working on!

**Comment by pfrazee** [comment, marked as answer]
Oof, sorry for the surprise behavior. I haven't exposed any UI for it, but you can also dig through the history of a drive to pull out deleted files. HMU if you ever need help recovering files.

So, there's two tricks with foldersync. The first is whether it should be bidirectional or unidirectional. Currently it's the latter - the source folder syncs into the drive, but the drive doesn't sync to the folder. Bidrectional is a little harder to suss out, but if that's what people expect, I think I can make it happen. I'll need to think a little bit more about the UI and behaviors for that.

The second trick is the mounts. Mounts add a bit of complexity... for instance, they can potentially go infinitely, so you can't just recursively sync all of them. If it's unidirectional sync, the answer is just to disable them from sync and they don't factor in. But if it's bidirectional sync, I figure I'll have to cook up a UI for configuring the behavior... ignore them? Sync one, but then what about the children? Etc.

Either way, this is just one of those features! I knew there'd be rough edges and it looks like this is the first. I'll need to invest more time in it.

For now, I think the solution is to first export the drive to your local folder. Then you can turn on autosync and (hopefully) it'll work as expected.

**Reply by willurmston** [reply]
Ah, I see. I get why syncing from folder -> drive would make sense if you want to use an external editor, but I definitely assumed it was bi-directional. Maybe it's the word "sync". I got the files back but for future reference, how do you revert to an older version of a drive?

RE mounts: maybe you could sync one level of mounts, or pick specific ones like the existing UI. Showing the file size could help people pick what to sync. It's unidirectional but Transmit does a good job of clearly communicating what will happen when you hit the button.
![image](https://user-images.githubusercontent.com/8092948/83841026-d8cf2680-a6b4-11ea-9f54-723c41ec6b43.png)

Another thing with mounts is that that you don't necessarily have write access to them, so what happens if you make a change to a read-only mount folder? Maybe you could prevent writes on that folder with OS file permissions? Just an idea.

Anyways thanks for your quick reply and for all your work. I've been able to make a near-complete p2p clone of a popular web service in a matter of days. The dev experience is great!

**Reply by pfrazee** [reply]
Greatly appreciate your detailed thoughts here. I think I can get all this working, it's just gonna take a little more dev.

So in hyper:// URLs, if you put a `+{num}` after the key but before the path, it'll pull up the data at the version it's specifying. Right now versions are just a "revision integer" but eventually we'll have tags. If you call `getInfo()` you can see the current version. So you just gotta iterate backwards until you find the file you deleted.

**Reply by pfrazee** [reply]
Example: `hyper://{key}+5/foo.txt`. Fifth revision

**Reply by willurmston** [reply]
Great! bi-directional sync will open up a lot of cool use cases.

**Reply by pfrazee** [reply]
Bi-directional sync as an option is still on the table, but for now I went with something a little simpler, see https://github.com/beakerbrowser/beaker/issues/1724#issuecomment-643070106
